services/statestorage/service.go

```diff
@@ -137,15 +144,39 @@ func (s *service) GetStateHash(ctx context.Context, input *services.GetStateHash
 		return nil, errors.Wrapf(err, "unsupported block height: block %v is not yet committed", input.BlockHeight)
 	}
 
-	value, err := s.persistence.ReadMerkleRoot(input.BlockHeight)
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	if input.BlockHeight+primitives.BlockHeight(s.config.StateStorageHistoryRetentionDistance()) <= s.height {
+		return nil, errors.Errorf("unsupported block height: block %v too old. currently at %v. keeping %v back", input.BlockHeight, s.height, primitives.BlockHeight(s.config.StateStorageHistoryRetentionDistance()))
+	}
+
+	value, err := s.layeredState.ReadStateHash(input.BlockHeight)
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not merkle root for block height %d", input.BlockHeight)
+		return nil, errors.Wrapf(err, "could not find a merkle root for block height %d", input.BlockHeight)
 	}
 	output := &services.GetStateHashOutput{StateRootHash: value}
 
 	return output, nil
 }
 
+func inflateChainState(csd []*protocol.ContractStateDiff) adapter.ChainState {
```

Comment:
why do you need this method? you have what seems to be a very similar method inside the `layeredState`

Reply:
there was at one point but i dont think its there anymore..